src/Model.php

```diff
@@ -1281,44 +1281,44 @@ public function newInstance(string $class = null, array $options = [])
      * If you wish to fully copy the data from one
      * model to another you should use:
      *
-     * $m->withPersistence($p2, false)->set($m->get())->save();
+     * $model->withPersistence($p2, false)->set($model->get())->save();
      *
-     * See https://github.com/atk4/data/issues/111 for
-     * use-case examples.
+     * See https://github.com/atk4/data/issues/111 for use-case examples.
      *
-     * @param Persistence $persistence
-     * @param mixed       $id
-     * @param string      $class
+     * @param mixed $id
      *
-     * @return $this
+     * @return static
      */
-    public function withPersistence($persistence, $id = null, string $class = null)
+    public function withPersistence(Persistence $persistence, $id = null, string $class = null)
     {
-        if (!$persistence instanceof Persistence) {
-            throw (new Exception('Please supply valid persistence'))
-                ->addMoreInfo('arg', $persistence);
-        }
-
-        if (!$class) {
-            $class = static::class;
-        }
+        $class = $class ?? static::class;
 
-        $m = new $class($persistence);
+        $model = new $class($persistence, $this->table);
 
         if ($this->id_field) {
             if ($id === true) {
-                $m->id = $this->id;
-                $m->set($m->id_field, $this->get($this->id_field));
+                $model->id = $this->id;
+                $model->set($model->id_field, $this->get($this->id_field));
             } elseif ($id) {
-                $m->id = null; // record shouldn't exist yet
-                $m->set($m->id_field, $id);
+                $model->id = null; // record shouldn't exist yet
+                $model->set($model->id_field, $id);
             }
         }
 
-        $m->data = $this->data;
-        $m->dirty = $this->dirty;
+        // include any fields defined inline
+        foreach ($this->fields as $fieldName => $field) {
+            if (!$model->hasField($fieldName)) {
+                $model->addField($fieldName, clone $field);
+            }
+        }
 
-        return $m;
+        $model->data = $this->data;
```

Comment:
I think this method duplicates some other code.

Why is copying data prop better than `->set($model->get())` (some mention in the title?)

The code before it is a duplicate of `asModel()`, right?

Reply:
No. `asModel` and `newInstance` use the same persistence.

>Why is copying data prop better than ->set($model->get()) (some mention in the title?)
I will remove this from the code comments as data is copied from the source model already.
What you expect from this method is a model clone with different persistence.

Reply:
Hmm, lets merge this